tests/Doctrine/Tests/Annotations/Type/ArrayTypeTest.php

```diff
@@ -0,0 +1,222 @@
+use ArrayObject;
+use Countable;
+use Doctrine\Annotations\Type\ArrayType;
+use Doctrine\Annotations\Type\BooleanType;
+use Doctrine\Annotations\Type\IntegerType;
+use Doctrine\Annotations\Type\IntersectionType;
+use Doctrine\Annotations\Type\MixedType;
+use Doctrine\Annotations\Type\ObjectType;
+use Doctrine\Annotations\Type\StringType;
+use Doctrine\Annotations\Type\UnionType;
+use IteratorAggregate;
+use PHPUnit\Framework\TestCase;
+use stdClass;
+use Traversable;
+
+final class ArrayTypeTest extends TestCase
+{
+    /**
+     * @dataProvider descriptionProvider()
+     */
+    public function testDescription(ArrayType $type, string $description) : void
+    {
+        self::assertSame($description, $type->describe());
+    }
+
+    /**
+     * @param mixed[] $value
+     *
+     * @dataProvider validValuesProvider()
+     */
+    public function testValidValues(ArrayType $type, array $value) : void
+    {
+        self::assertTrue($type->validate($value));
+    }
+
+    /**
+     * @param mixed $value
+     *
+     * @dataProvider invalidValuesProvider()
+     */
+    public function testInvalidValues(ArrayType $type, $value) : void
+    {
+        self::assertFalse($type->validate($value));
+    }
+
+    public function testGetKeyType() : void
+    {
+        $type = new ArrayType(new IntegerType(), new StringType());
+
+        self::assertInstanceOf(IntegerType::class, $type->getKeyType());
+    }
+
+    public function testGetValueType() : void
+    {
+        $type = new ArrayType(new IntegerType(), new StringType());
+
+        self::assertInstanceOf(StringType::class, $type->getValueType());
+    }
+
+    /**
+     * @return (ArrayType|string)[]
+     */
+    public function descriptionProvider() : iterable
+    {
+        yield [
+            new ArrayType(new MixedType(), new MixedType()),
+            'array<mixed>',
+        ];
+        yield [
+            new ArrayType(new MixedType(), new IntegerType()),
+            'array<int>',
+        ];
+        yield [
+            new ArrayType(new UnionType(new IntegerType(), new StringType()), new IntegerType()),
+            'array<int>',
+        ];
+        yield [
+            new ArrayType(new UnionType(new StringType(), new IntegerType()), new IntegerType()),
+            'array<int>',
+        ];
+        yield [
+            new ArrayType(new IntegerType(), new IntegerType()),
+            'array<int, int>',
+        ];
+        yield [
+            new ArrayType(new IntegerType(), new UnionType(new IntegerType(), new BooleanType())),
+            'array<int, int|bool>',
+        ];
+        yield [
+            new ArrayType(
+                new UnionType(new IntegerType(), new ObjectType(null)),
+                new UnionType(
+                    new StringType(),
+                    new IntersectionType(new ObjectType(Traversable::class), new ObjectType(Countable::class))
+                )
+            ),
+            'array<int|object, string|(Traversable&Countable)>',
```

Comment:
should we really allow to have possible `object` as array key?

Reply:
As it is now, yes. Whether it works in parser/language is another thing.
PHPStan does that also btw: https://github.com/phpstan/phpstan/blob/master/src/Type/ArrayType.php

Reply:
> As it is now, yes.

Why?

Reply:
Explained above in a comment for ArrayType.